Mathlib/Analysis/Meromorphic/FactorizedRational.lean

```diff
@@ -0,0 +1,147 @@
+  ext u
+  constructor <;> intro h
+  · simp_all only [mem_mulSupport, ne_eq, mem_support]
+    by_contra hCon
+    simp_all [hCon, zpow_zero]
+  · simp_all only [mem_mulSupport, ne_eq, ne_iff]
+    use u
+    simp_all [zero_zpow_eq_one₀]
+
+/--
+Factorized rational functions are analytic wherever the exponent is non-negative.
+-/
+theorem analyticAt {d : 𝕜 → ℤ} {x : 𝕜} (h : 0 ≤ d x) :
+    AnalyticAt 𝕜 (∏ᶠ u, (· - u) ^ d u) x := by
+  apply analyticAt_finprod
+  intro u
+  by_cases h₂ : x = u
+  · apply AnalyticAt.fun_zpow_nonneg (by fun_prop)
+    rwa [← h₂]
+  · apply AnalyticAt.fun_zpow (by fun_prop)
+    rwa [sub_ne_zero]
+
+/--
+Factorized rational functions are non-zero wherever the exponent is zero.
+-/
+theorem ne_zero {d : 𝕜 → ℤ} {x : 𝕜} (h : d x = 0) :
+    (∏ᶠ u, (· - u) ^ d u) x ≠ 0 := by
+  by_cases h₁ : (fun u ↦ (· - u) ^ d u).mulSupport.Finite
+  · rw [finprod_eq_prod _ h₁, Finset.prod_apply, Finset.prod_ne_zero_iff]
+    intro z hz
+    simp only [Pi.pow_apply, ne_eq]
+    by_cases h₂ : x = z <;> simp_all [zpow_ne_zero, sub_ne_zero]
+  · simp [finprod_of_infinite_mulSupport h₁]
+
+
+open Classical in
+/--
+Helper Lemma for Computations: Extract one factor out of a factorized rational function.
+-/
+lemma extractFactor {d : 𝕜 → ℤ} (u₀ : 𝕜) (hd : d.support.Finite) :
+    (∏ᶠ u, (· - u) ^ d u) = ((· - u₀) ^ d u₀) * (∏ᶠ u, (· - u) ^ (update d u₀ 0 u)) := by
+  by_cases h₁d : d u₀ = 0
+  · simp [← eq_update_self_iff.2 h₁d, h₁d]
+  · have : (fun u ↦ (fun x ↦ x - u) ^ d u).mulSupport ⊆ hd.toFinset := by
+      simp [mulSupport]
+    rw [finprod_eq_prod_of_mulSupport_subset _ this]
+    have : u₀ ∈ hd.toFinset := by simp_all
+    rw [← Finset.mul_prod_erase hd.toFinset _ this]
+    congr 1
+    have : (fun u ↦ (· - u) ^ (update d u₀ 0 u)).mulSupport ⊆ hd.toFinset.erase u₀ := by
+      rw [mulSupport]
+      intro x hx
+      by_cases h₁x : x = u₀ <;> simp_all
+    simp_all [finprod_eq_prod_of_mulSupport_subset _ this, Finset.prod_congr rfl]
+
+/--
+Factorized rational functions are meromorphic in normal form on `Set.univ`.
+-/
+theorem meromorphicNFOn_univ (d : 𝕜 → ℤ) :
+    MeromorphicNFOn (∏ᶠ u, (· - u) ^ d u) Set.univ := by
+  classical
+  by_cases hd : d.support.Finite
+  · intro z hz
+    rw [extractFactor z hd]
+    right
+    use d z, (∏ᶠ u, (· - u) ^ update d z 0 u)
+    simp [analyticAt, ne_zero]
+  · rw [← mulSupport d] at hd
+    rw [finprod_of_infinite_mulSupport hd]
+    exact AnalyticOnNhd.meromorphicNFOn analyticOnNhd_const
+
+/--
+Factorized rational functions are meromorphic in normal form on arbitrary subsets of `𝕜`.
+-/
+theorem meromorphicNFOn (d : 𝕜 → ℤ) (U : Set 𝕜) :
+    MeromorphicNFOn (∏ᶠ u, (· - u) ^ d u) U := fun _ _ ↦ meromorphicNFOn_univ d (trivial)
+
+/--
+The order of the factorized rational function `(∏ᶠ u, fun z ↦ (z - u) ^ d u)` at `z` equals `d z`.
+-/
+theorem order {z : 𝕜} (d : 𝕜 → ℤ) (h₁d : d.support.Finite) :
+    (meromorphicNFOn_univ d (trivial : z ∈ ⊤)).meromorphicAt.order = d z := by
```

Comment:
```suggestion
    (meromorphicNFOn_univ d (mem_univ z)).meromorphicAt.order = d z := by
```
You will need to `open Set` at the beginning of the file for this.

Same thing below.

Reply:
Implemented.